eth/tracers/firehose_concurrency_test.go

```diff
@@ -0,0 +1,170 @@
+		private: &privateFirehoseConfig{
+			FlushToTestBuffer: true,
+		},
+	})
+
+	f.OnBlockchainInit(params.AllEthashProtocolChanges)
+
+	blockHashes := make(map[uint64]string, blockCount)
+
+	for i := 0; i < blockCount; i++ {
+		blockNum := uint64(baseBlockNum + i)
+
+		f.OnBlockStart(blockEvent(blockNum))
+		blockHashes[blockNum] = hex.EncodeToString(f.block.Hash) // Store hash before block reset
+
+		f.onTxStart(txEvent(), hex2Hash(fmt.Sprintf("TX%d", i)), from, to)
+		f.OnCallEnter(0, byte(vm.CALL), from, to, nil, 0, nil)
+		f.OnBalanceChange(from, b(100), b(50), 0)
+		f.OnCallExit(0, nil, 0, nil, false)
+		f.OnTxEnd(txReceiptEvent(0), nil)
+
+		f.OnBlockEnd(nil)
+	}
+
+	f.OnClose()
+
+	output := f.InternalTestingBuffer().String()
+	extractedBlocks := extractBlocksFromOutput(t, output)
+
+	// Verify block count
+	require.Equal(t, blockCount, len(extractedBlocks),
+		"Expected %d blocks in output, found %d", blockCount, len(extractedBlocks))
+
+	// Verify blocks in order
+	verifyBlockSequence(t, extractedBlocks, baseBlockNum)
+
+	// Verify block hashes
+	for _, block := range extractedBlocks {
+		expectedHash, exists := blockHashes[block.number]
+		require.True(t, exists, "Block %d not found in tracked blocks", block.number)
+		require.Equal(t, expectedHash, block.hash,
+			"Hash mismatch for block %d", block.number)
+	}
+}
+
+type extractedBlock struct {
+	number uint64
+	hash   string
+}
+
+func extractBlocksFromOutput(t *testing.T, output string) []extractedBlock {
+	t.Helper()
+
+	// Regex to extract the block number and hash from the FIRE BLOCK line
+	blockInfoRegex := regexp.MustCompile(`FIRE BLOCK (\d+) ([0-9a-fA-F]+)`)
+
+	lines := strings.Split(output, "\n")
+	var blocks []extractedBlock
+
+	for _, line := range lines {
+		if strings.HasPrefix(line, "FIRE BLOCK") {
+			matches := blockInfoRegex.FindStringSubmatch(line)
+			if len(matches) == 3 {
+				blockNumStr := matches[1]
+				blockHash := matches[2]
+
+				blockNum, err := strconv.ParseUint(blockNumStr, 10, 64)
+				require.NoError(t, err, "failed to parse block number: %s", blockNumStr)
+
+				blocks = append(blocks, extractedBlock{
+					number: blockNum,
+					hash:   blockHash,
+				})
+			}
+		}
+	}
+
+	return blocks
+}
+
+func verifyBlockSequence(t *testing.T, blocks []extractedBlock, baseBlockNum uint64) {
+	t.Helper()
+
+	// First block should be the base block number
+	require.Equal(t, baseBlockNum, blocks[0].number,
+		"First block should be %d, got %d", baseBlockNum, blocks[0].number)
+
+	// Last block should be base + count - 1
+	expectedLast := baseBlockNum + uint64(len(blocks)) - 1
+	require.Equal(t, expectedLast, blocks[len(blocks)-1].number,
+		"Last block should be %d, got %d", expectedLast, blocks[len(blocks)-1].number)
+
+	// Verify sequence
+	for i := 0; i < len(blocks)-1; i++ {
+		current := blocks[i].number
+		next := blocks[i+1].number
+		require.Equal(t, current+1, next,
+			"Blocks out of order at position %d: %d followed by %d", i, current, next)
+	}
+}
```

Comment:
```suggestion
	for i, block := range blocks {
		require.Equal(t, baseBlockNum+uint64(i), block.number, "Blocks out of order at position %d", i)
	}
```

This is equivalent in terms of testing as previous code, I think there is too much complexity in your initial version.  With this reduced, no need for a `verifyBlockSequence` anymore, can be inlined where called.